### DISCUSSION TITLE
7 DOF Beam

### DISCUSSION THREAD

**Discussion by andrey1887** [idaholab/moose#25754] [Q&A Modules: Solid mechanics]
Hello. I want to try to implement line element with 7 degree of freedom in each node. 7-th degree will consider warping of cross-section. Besides of this, bimoment must be calculated. I have a stiffness matrix of this element in explicit form. Does existing capabilities of MOOSE can help to discribe this problem ? Thank you for suggestions.
![MicrosoftTeams-image](https://github.com/idaholab/moose/assets/114737834/ca8fad3f-5766-4031-845b-36c935bcc863)
![MicrosoftTeams-image (1)](https://github.com/idaholab/moose/assets/114737834/bb597cd5-f6a9-49b7-b597-4fc6099e3a39)
![MicrosoftTeams-image (2)](https://github.com/idaholab/moose/assets/114737834/909dd139-2294-418e-bbcb-34f7b644b2bb)
oose/assets/114737834/855d0cfc-6071-413e-8a02-c8b5f4d6981c)

**Comment by GiudGiud** [comment]
@cbolisetti @roystgnr @dschwen 

The answer is no, but it could be developed, in libmesh

**Reply by andrey1887** [reply]
Thank you for answer. Will you take a some link to information or example how to do it. 
There is supposed subsequent interaction between MOOSE and libmesh isn't it? Is MOOSE still used if we use libmesh?
@cbolisetti @roystgnr @dschwen @GiudGiud

**Reply by GiudGiud** [reply]
You can use libmesh without moose.

However moose provides you with a friendlier setup, post processing capabilities and a lot of pre existing equation implementations.

**Reply by andrey1887** [reply]
@GiudGiud do you have some example of using libmesh with moose ?

**Reply by GiudGiud** [reply]
All of moose relies on libmesh as the backend.

**Reply by GiudGiud** [reply]
All of moose relies on libmesh as the backend.

**Reply by GiudGiud** [reply]
All of moose relies on libmesh as the backend.

**Reply by roystgnr** [reply]
You can get 7 degrees of freedom per node (on a 2-node element, I assume?  "all line elements have 2 nodes" isn't true) by adding seven (default, `LAGRANGE`) variables to your system.  After that?  If your stiffness matrix comes from an integral form, you can write a new `Kernel` object for it and overload `computeQpResidual()` or `computeQpJacobian()`; if it doesn't have an integral representation then you can overload `computeResidual()` and `computeJacobian()` instead.

**Reply by andrey1887** [reply]
@roystgnr , Thank you for answer. So, do I need additional development in libmesh?

**Reply by cbolisetti** [reply]
I think you can do this in MOOSE without libmesh development. The current beam element implementation is almost completely in MOOSE and doesn't use much libmesh (@bwspenc please correct me if I am wrong). There are a couple of options here: 

- You could do a quick and dirty implementation like @roystgnr described (for the case where the stiffness matrix is not in an integral form, which is your scenario, I think). We already have a template you can start with in MASTODON in the material called `LinearSpring`, which can be used with line elements. This material implements a simple, diagonal stiffness matrix for a 6-dof line element. This can be extended to the stiffness matrix you provided above. I actually recommend this method to begin with. (Note that in MOOSE, we create the mesh with line elements, assign appropriate number of DOFs, and then use the `Materials` to model it as a beam, truss, or a spring. This is different from traditional structural engineering software.)
- You could also extend the current beam element implementation in `ComputeBeamElasticity` and `StressDivergenceBeam` objects. This is less straightforward, since this implementation is based on a 6-dof assumption.

**Reply by andrey1887** [reply]
Thank you, @cbolisetti. Is this method (using MASTODON) suitable for the case when matrix elements are specified explicity?
If yes, then do i need to install BlackBear with it?

**Reply by cbolisetti** [reply]
You still have to develop a new material for your stiffness matrix with the 7 dof element. I don't think there is anything in MOOSE or MASTODON that can do what you are describing without extra coding. The `LinearSpring` material merely provides a starting point that you can use if you want to override the `computeResidual()` and `computeJacobian()` methods as described by @roystgnr. The starting point is for a simple diagonal stiffness matrix with constant properties, but you will have to code and develop a new material that uses a stiffness matrix you provided above.

**Reply by andrey1887** [reply]
Okay. I will try it. Thank you.

**Reply by andrey1887** [reply]
@cbolisetti,  as i undestand, MASTODON is a moose based app so I need to work within it and not create my own app. Can i copy some snippets from LinearSpring.C to my app ?

**Reply by cbolisetti** [reply]
Hi @andrey1887, yes, you can copy code snippets. Also, you don't have to put this in mastodon. Since the application of this element may go beyond just seismic analysis, you can put it in the tensormechanics module in MOOSE. 
You can still copy code snippets from LinearSpring and follow the general structure. You'll have to extend it to calculate deformation for the 7th dof and change the stiffness matrix according to the beam equations.

**Reply by andrey1887** [reply]
Thank you, @cbolisetti . As you mentioned, i need to override `computeResidual()` and `computeJacobian()` methods, but i didn't find them in LinearSpring.C may be they are defined in another place?

**Reply by cbolisetti** [reply]
That happens in `StressDivergenceSpring`, which is the corresponding Kernel object. The `LinearSpring` material only provides the input (forces and deformations) for calculating the residual in `StressDivergenceSpring`.

**Reply by andrey1887** [reply]
@cbolisetti , @GiudGiud  Besides `computeResidual()` and `computeJacobian() ` there is a method `computeStiffnessMatrix()`,
here stiffness matrix consits of the next components:
The stiffness matrix is of the form
  |  kdd  kdr  |
  |  krd  krr  |
  where kdd, krr, krd and kdr are all RankTwoTensors (3x3 matrix) and
  matrix symmetry is assumed, namely, krd = kdr.transpose()
  This implementation of the spring element has only diagonal stiffness
  terms. Therefore, Kdr and Krd are zero. 
  
  Or for   `Compute incremental beam Strain`

`K = |K11 K12|
        |K21 K22|`
        
     

And the matrix dimension is 6x6 
How to divide my matrix (14x14) on appropriate parts :

![MicrosoftTeams-image (5)](https://github.com/idaholab/moose/assets/114737834/4be14a28-4743-464f-bce0-814a8cc6ef19)
![MicrosoftTeams-image (3)](https://github.com/idaholab/moose/assets/114737834/bd31ab6e-253d-4d19-979c-3f61bc9adfb4)

**Reply by GiudGiud** [reply]
That's not really my expertise. @cbolisetti @bwspenc is there a reference for doing that in MOOSE?

**Reply by cbolisetti** [reply]
It looks like this matrix is giving you forces at individual nodes? Is that correct? If you want to use the simple spring like formulation, you'd want a stiffness matrix that gives you forces in the element, not on the individual nodes. That said you might be able to convert this matrix to a 7x7 that does F=KU, where U is the relative deformation between the two nodes.

**Reply by andrey1887** [reply]
Thank you for your answers. @cbolisetti, what do you think, how this conversion can be done? And how this affect the results of analysis. I saw the  `ComputeIncrementalBeamStrain.C` but it's `computeStiffnessMatrix` method creates a 6x6 matrix too (as i understand), is it right? [https://github.com/idaholab/moose/blob/0721a91922831209ae46f28d875a48c3e84b3f2c/modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C#L345](url)

**Reply by andrey1887** [reply]
Hello. @cbolisetti, what do tou think about using `ComputeBeamElasticity` and `StressDivergenceBeam`.
Now i want to gather all the kernels and materials for my task. The forces and moments calculation is supposed too.
I think it will be looks like:

Kernels:
1. Stress Divergence Beam

Materials:
1. ComputeIncrementalBeamStrain
2. ComputeElasticityBeam
3. ComputeBeamResultants 

Is it enough?

**Comment by andrey1887** [comment]
Help me please. Is it possible to use use stiffness matrix in this form ![MicrosoftTeams-image (3)](https://github.com/idaholab/moose/assets/114737834/bd31ab6e-253d-4d19-979c-3f61bc9adfb4)
instead of 6x6 matrix defined in:
```
ComputeIncrementalBeamStrain::computeStiffnessMatrix()
{
  const Real youngs_modulus = _material_stiffness[0](0);
  const Real shear_modulus = _material_stiffness[0](1);

  const Real A_avg = (_area[0] + _area[1]) / 2.0;
  const Real Iy_avg = (_Iy[0] + _Iy[1]) / 2.0;
  const Real Iz_avg = (_Iz[0] + _Iz[1]) / 2.0;
  Real Ix_avg = (_Ix[0] + _Ix[1]) / 2.0;
  if (!_has_Ix)
    Ix_avg = Iy_avg + Iz_avg;

  // K = |K11 K12|
  //     |K21 K22|

  // relation between translational displacements at node 0 and translational forces at node 0
  RankTwoTensor K11_local;
  K11_local.zero();
  K11_local(0, 0) = youngs_modulus * A_avg / _original_length[0];
  K11_local(1, 1) = shear_modulus * A_avg / _original_length[0];
  K11_local(2, 2) = shear_modulus * A_avg / _original_length[0];
  _K11[0] = _total_rotation[0].transpose() * K11_local * _total_rotation[0];

  // relation between displacements at node 0 and rotational moments at node 0
  RankTwoTensor K21_local;
  K21_local.zero();
  K21_local(2, 1) = shear_modulus * A_avg * 0.5;
  K21_local(1, 2) = -shear_modulus * A_avg * 0.5;
  _K21[0] = _total_rotation[0].transpose() * K21_local * _total_rotation[0];

  // relation between rotations at node 0 and rotational moments at node 0
  RankTwoTensor K22_local;
  K22_local.zero();
  K22_local(0, 0) = shear_modulus * Ix_avg / _original_length[0];
  K22_local(1, 1) = youngs_modulus * Iz_avg / _original_length[0] +
                    shear_modulus * A_avg * _original_length[0] / 4.0;
  K22_local(2, 2) = youngs_modulus * Iy_avg / _original_length[0] +
                    shear_modulus * A_avg * _original_length[0] / 4.0;
  _K22[0] = _total_rotation[0].transpose() * K22_local * _total_rotation[0];

  // relation between rotations at node 0 and rotational moments at node 1
  RankTwoTensor K22_local_cross = -K22_local;
  K22_local_cross(1, 1) += 2.0 * shear_modulus * A_avg * _original_length[0] / 4.0;
  K22_local_cross(2, 2) += 2.0 * shear_modulus * A_avg * _original_length[0] / 4.0;
  _K22_cross[0] = _total_rotation[0].transpose() * K22_local_cross * _total_rotation[0];

  // relation between displacements at node 0 and rotational moments at node 1
  _K21_cross[0] = -_K21[0];
```
